### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by zhamid01**
Hope everything is good!

### CODE REVIEW

File: src/main/java/Problem1.java [hunk 1]
```diff
@@ -1,10 +1,41 @@
...
+                first++;
+                middle++;
+            }
+            else if (nums[middle] == 1) {
+                middle++;
+            }
+            else if (nums[middle] == 2) {
+                holder = nums[middle];
+                nums[middle] = nums[last];
+                nums[last] = holder;
```
Comment (pdgetrf) [inline src/main/java/Problem1.java:36]: I see repeated logic here for swap. refactor it into a private function. we prefer single source of truth in our code.

File: src/main/java/Problem2.java [hunk 2]
```diff
@@ -1,12 +1,32 @@
...
+    */
 
     // Do not change signature (function name, parameters and return type)
     // Do not use extra space for another array. Solution to be in-pace with O(1) extra memory
     public static int remove(int[] nums) {
-        return Integer.MAX_VALUE; // place holder
+        if (nums.length - 1 == 0) {
+            return 1;
+        }
+        int tempArray[] = new int[nums.length];
```
Comment (pdgetrf) [inline src/main/java/Problem2.java:19]: I think you missed this requirement:

```
This is to be done in-place which means any extra space for another array is not allowed. 
```

File: src/main/java/Problem3/Book.java [hunk 3]
```diff
@@ -15,11 +15,17 @@ public Book(String title, String author) {
...
     }
 
     @Override
     public boolean equals(Object obj) {
-        // homework
+        Book other = (Book) obj;
+        if (other == null) {
+            System.out.println("Error: null Book object");
+        }
+        return (this.id == other.id && this.title == other.title && this.author == other.author);
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:29]: wrong logic here

```
A movie equals another movie only when their id are the same. Same rule for books being equal.
```

File: src/main/java/Problem3/BookRomance.java [hunk 4]
```diff
@@ -5,13 +5,15 @@
     private int lateFeePerDayInDollar = 4;
 
     public BookRomance(String title, String author) {
-        // homework
-        // tip: use the 'super' keyword
+        super (title, author);
+        this.title = title;
+        this.author = author;
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookRomance.java:10]: line 9 and 10 not needed after line 8

File: src/main/java/Problem3/Movie.java [hunk 5]
```diff
@@ -9,15 +9,22 @@
...
     }
 
     @Override
     public boolean equals(Object obj) {
-        // homework
+        Movie other = (Movie) obj;
+        if (other == null) {
+            System.out.println("Error: null Book object");
+        }
+        return (this.id == other.id && this.title == other.title && this.rating == other.rating);
```
Comment (pdgetrf) [inline src/main/java/Problem3/Movie.java:28]: same here. wrong logic. only compare id.

File: src/main/java/Problem3/MovieAction.java [hunk 6]
```diff
@@ -21,6 +23,12 @@ public int getLateFeeInDollar() {
...
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
-        // homework
+        if (numOfDaysPastDue > 0 && numOfDaysPastDue < 5) {
+            return (numOfDaysPastDue * lateFeePerDayInDollar);
+        }
+        else if (numOfDaysPastDue >= 5){
+            return ((2 * numOfDaysPastDue) * lateFeePerDayInDollar);
+        }
+        return 0;
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:32]: is this line reachable?

Comment (pdgetrf) [reply src/main/java/Problem3/MovieAction.java:32]: also duplicated logic here. 

suggesting:

```java
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

File: src/test/java/Problem2Test.java [hunk 7]
```diff
@@ -7,13 +7,19 @@
...
         };
 
         int[][] expects = {
-                {0, 1, 2, 4, 6}
-                // add more test cases
+                {0, 1, 2, 4, 6},
+                {0, 1, 2, 4, 5, 6},
+                {1, 2, 4, 6},
+                {0, 1, 2, 3, 4, 5},
+                {1, 2, 3, 4, 5}
```
Comment (pdgetrf) [inline src/test/java/Problem2Test.java:22]: good test.

need more edge cases like null array, empty array, single element
